### PR TITLE
Implemented new configuration API

### PR DESCRIPTION
**Pull request by thecodejunkie**
See #1999 for more details on the contents of this post

### CODE REVIEW

**Comment by jchannon** [comment]
:shipit:

File: src/Nancy.Tests/Unit/Bootstrapper/NancyBootstrapperBaseFixture.cs [hunk 1]
```diff
@@ -465,6 +475,14 @@ public void Dispose()
         public bool Disposed { get; private set; }
     }
 
+    internal class FakeEnvironmentalConfigurator : INancyEnvironmentConfigurator
```
Comment (horsdal) [inline src/Nancy.Tests/Unit/Bootstrapper/NancyBootstrapperBaseFixture.cs]: `Environmental` should just be `Environment`

File: src/Nancy.Tests/Unit/Configuration/DefaultNancyEnvironmentFixture.cs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+        [Theory]
+        [InlineData(2)]
+        [InlineData(4)]
+        [InlineData(6)]
+        public void Should_add_value_to_environment_when_invoking_addvalue(int numberOfElementsToAdd)
+        {
+            // Given, When
+            var environment = CreateEnvironment(numberOfElementsToAdd);
+            IReadOnlyDictionary<string, object> setter = environment;
```
Comment (horsdal) [inline src/Nancy.Tests/Unit/Configuration/DefaultNancyEnvironmentFixture.cs]: No reason to have the `setter` variable` AFAICT

File: src/Nancy.Tests/Unit/Configuration/DefaultNancyEnvironmentFixture.cs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+            // When
+            var result = environment.ContainsKey(keyToLookFor);
+
+            // Then
+            result.ShouldEqual(expectedResult);
+        }
+
+        [Fact]
+        public void Should_return_true_if_key_exists_when_invoking_trygetvalue()
```
Comment (horsdal) [inline src/Nancy.Tests/Unit/Configuration/DefaultNancyEnvironmentFixture.cs]: This test and `Should_return_false_if_key_does_not_exist_when_invoking_trygetvalue()` could be unified with some `InilineData` like above.

**Comment by horsdal** [comment]
How about `INacyEnvironment` implements 2 interfaces. A mutable one and an immutable one. The environment is passed as the mutable interface into `.Configure`, but only registered in the container as the immutable one. Defaults could then be overriden in `Configure` but module etc. taking a dependency on the environment could only read.

**Comment by thecodejunkie** [comment]
Thanks @horsdal updated the code according to your review comments!

**Comment by phillip-haydon** [comment]
:+1:

**Comment by thecodejunkie** [comment]
Rebased against latest `v2` branch. We need to pull this before we merge `v2` into `master` .. same with #2012 after I've rebased it

**Pull request merged**